src/test/java/org/junitpioneer/jupiter/resource/PathDeleterTests.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2016-2022 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter.resource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+
+import java.io.IOException;
+import java.nio.file.FileSystem;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import com.google.common.jimfs.Configuration;
+import com.google.common.jimfs.Jimfs;
+
+import org.junit.jupiter.api.Test;
+
+class PathDeleterTests {
```

Comment:
I just noticed that I myself added tests here, but before that I wanted to comment with:

> I understand we're having fun writing tests, but so much fun that we're testing other peoples' code? :) Or are you testing assumptions of how these APIs work?

😁

Reply:
I was test-driving my code the top-down "mockist" fashion, and when I eventually got to testing this deep in the call stack, I realised that I needed to test the "unhappy paths" of the parts of the code using `Files::walkFileTree`. Specifically, what happens if the temporary files were somehow deleted before `Files::walkFileTree` runs?

The easy solution is to use `Files::delete`, but that would throw throw `IOException`s if the files were deleted already, and I wanted to the code to silently do nothing instead. So I wanted to use `Files::deleteIfExists`. But how can one test that? The on-disk filesystem is inherently so non-deterministic that I couldn't think how to consistently delete a file created by `TemporaryDirectory` before it got to the `Files::walkFileTree` code.

So I split the code passed to `Files::walkFileTree` into its own unit, the `PathDeleter`, and tested that independently.

I hope this makes sense. :grin:

Reply:
I'm an idiot, I just overlooked the use of `PathDeleter.INSTANCE`. 🤦🏾 Even after reading your reply, it took me a few minutes.

Reply:
No worries, my explanation was quite convoluted, so I don't blame you for taking a few minutes. :)